[RF-DOCS] Action View Overview Guide [ci-skip] (#51435)
rails/rails: guides/source/action_view_overview.md

> **DO NOT READ THIS FILE ON GITHUB, GUIDES ARE PUBLISHED ON
> https://guides.rubyonrails.org.**

Reviewer: Let's revert this first line to not break.

```suggestion
**DO NOT READ THIS FILE ON GITHUB, GUIDES ARE PUBLISHED ON https://guides.rubyonrails.org.**
```

What is Action View?
--------------------

Action View is the V in [MVC](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller). [Action Controller](action_controller_overview.html) and Action View work together to handle web requests. Action Controller is concerned with communicating with the model layer (of MVC) and retrieving data. Action View is then responsible for compiling a response to the web request using that data.

> Action View templates (also referred to simply as "views") are written using Embedded Ruby (ERB), which allows using Ruby code within HTML documents.

Reviewer 2: Good point @MatheusRich , for a new rubyst should be interesting to know how flexible the framework is.

> Action View templates (also referred to simply as "views") are written using Embedded Ruby (ERB), which allows using Ruby code within HTML documents.

Reviewer 3: While this is true, I don't think it'ts good to encourage the use of Haml or Slim in the official guides. Maybe mention json or xml only?

> Action View templates (also referred to simply as "views") are written using Embedded Ruby (ERB), which allows using Ruby code within HTML documents.

Reviewer: We mention this under "Templates": `Other libraries may add other template types and file extensions as well.`. It's okay to say that support can be extended by external libraries, but I don't think we need to be specific with Haml/Slim, because tomorrow another one might show up and we'd need to update, etc.

> Action View templates (also referred to simply as "views") are written using Embedded Ruby (ERB), which allows using Ruby code within HTML documents.

Reviewer 4: I think the updated text is enough now. No need to cite explicit libraries here.

> Action View provides many [helper](#helpers) classes for dynamically generating HTML tags for forms, dates, and strings. It's also possible to add custom helpers to your application as needed.

Reviewer 3: It should also be [helpers] instead of [helper].

> Action View provides many [helper](#helpers) classes for dynamically generating HTML tags for forms, dates, and strings. It's also possible to add custom helpers to your application as needed.

Reviewer 5: I think it should be `[helper methods](#helpers)` or just `[helpers](#helpers)`. "helpers methods" doesn't sound right.

> NOTE: Some features of Action View are tied to Active Model, but that doesn't mean Action View depends on Active Model. Action View is an independent package that can be used with any Ruby library.

Reviewer 4: Should we link to one of those features or put an example here?

Reviewer 6: @bhumi1102 @MatheusRich Maybe link to `to_param` and `to_partial_path`?

```suggestion
NOTE: Action View can make use of Active Model features like [`to_param`](https://api.rubyonrails.org/classes/ActiveModel/Conversion.html#method-i-to_param) and [`to_partial_path`](https://api.rubyonrails.org/classes/ActiveModel/Conversion.html#method-i-to_partial_path) to simplify code. That doesn't mean Action View depends on Active Model. Action View is an independent package that can be used with any Ruby library.
```

      [...]
```

The file names follow a Rails naming convention. They share their name with the associated controller action. For example the `index.html.erb`, `edit.html.erb`, etc. above.

> By following this naming convention, Rails will automatically find and render the matching view at the end of a controller action, without you having to  specify it. For example, the `index` action in the `articles_controller.rb` will automatically render the `index.html.erb` view inside the `app/views/articles/` directory. The name and the location of the file are both important.

Reviewer 5: Could we have an example with a nested controller too?

The final HTML returned to the client is composed of a combination of the `.html.erb` ERB file, a layout template that wraps it, and all the partials that the ERB file may reference. In the rest of this guide, you will find more details about each of the three components: `Templates`, `Partials`, `Layouts`.

Templates
---------

Action View templates can be written in different formats. If the template file has a `.erb` extension, it uses embedded Ruby to build an HTML response. If the template has a `.jbuilder` extension, it uses the [Jbuilder](https://github.com/rails/jbuilder) gem to build a JSON response. And a template with a `.builder` extension uses the [`Builder::XmlMarkup`](https://github.com/jimweirich/builder) library to build an XML response.

Rails uses the file extension to distinguish among multiple template systems. For example, an HTML file using the ERB template system will have `.html.erb` as a file extension. Other libraries may add other template types and file extensions as well.

### ERB

> An ERB template is a way to sprinkle Ruby code within static HTML using special ERB tags `<% %>` and `<%= %>` tags.

Reviewer 5: It might be useful to link to the documentation for ERB syntax somewhere in this section so people can find the additional tags.

      [...]
```

> The file names follow a Rails naming convention. They share their name with the associated controller action. For example the `index.html.erb`, `edit.html.erb`, etc. above.

Reviewer 7: ```suggestion
The file names follow a Rails naming convention. They share their name with the associated controller action. For example the `index.html.erb`, `edit.html.erb`, etc.
```

By following this naming convention, Rails will automatically find and render the matching view at the end of a controller action, without you having to specify it. For example, the `index` action in the `articles_controller.rb` will automatically render the `index.html.erb` view inside the `app/views/articles/` directory. The name and the location of the file are both important.

The final HTML returned to the client is composed of a combination of the `.html.erb` ERB file, a layout template that wraps it, and all the partials that the ERB file may reference. In the rest of this guide, you will find more details about each of the three components: `Templates`, `Partials`, `Layouts`.

Templates
---------

> Action View templates can be written in different formats. If the template file has a `.erb` extension, it uses embedded Ruby to build an HTML response. If the template has a `.jbuilder` extension, it uses the [Jbuilder](https://github.com/rails/jbuilder) gem to build a JSON response. And a template with a `.builder` extension uses the [`Builder::XmlMarkup`](https://github.com/jimweirich/builder) library to build an XML response.

Reviewer 7: ```suggestion
Action View templates can be written in different formats. If the template file has a `.erb` extension, it uses embedded Ruby to build a HTML response. If the template has a `.jbuilder` extension, it uses the [Jbuilder](https://github.com/rails/jbuilder) gem to build a JSON response. And a template with a `.builder` extension uses the [`Builder::XmlMarkup`](https://github.com/jimweirich/builder) library to build a XML response.
```

Reviewer 6: "an HTML" and "an XML" are correct in this case because they are both pronounced with a vowel sound.

associated controller action. For example the `index.html.erb`, `edit.html.erb`,
etc.

By following this naming convention, Rails will automatically find and render
the matching view at the end of a controller action, without you having to
specify it. For example, the `index` action in the `articles_controller.rb` will
automatically render the `index.html.erb` view inside the `app/views/articles/`
directory. The name and the location of the file are both important.

The final HTML returned to the client is composed of a combination of the
`.html.erb` ERB file, a layout template that wraps it, and all the partials that
the ERB file may reference. In the rest of this guide, you will find more
details about each of the three components: `Templates`, `Partials`, `Layouts`.

Templates
---------

Action View templates can be written in different formats. If the template file
has a `.erb` extension, it uses embedded Ruby to build an HTML response. If the
template has a `.jbuilder` extension, it uses the
[Jbuilder](https://github.com/rails/jbuilder) gem to build a JSON response. And
a template with a `.builder` extension uses the

> [`Builder::XmlMarkup`](https://github.com/jimweirich/builder) library to build

Reviewer 4: This is a bit unrelated to this PR, but I wonder if we are really using Jim Weirich's version of this Gem. His repo hasn't been updated in 11 years (for obvious reasons), but [RubyGems says](https://rubygems.org/gems/builder/versions/3.2.4) the latest version is 3.2.4. Is this gem being maintained elsewhere?

Reviewer 6: It seems it's being maintained by Aaron: https://github.com/tenderlove/builder who is also mentioned as an [owner](https://rubygems.org/gems/builder) on Rubygems.

Reviewer 4: Should we link to that repo instead? I'm confused as to why RubyGems points to Jim's repo. Aaron's repo seem to have the right info on gemspec.

Reviewer 3: Should this even be mentioned in the official guides? Since it's not really a gem maintained by the Rails team as such.

Reviewer 6: The homepage of tenderlove's version points to tenderlove's version.
https://github.com/tenderlove/builder/blob/b3941919becdce4021a3f70838240442f155f268/builder.gemspec#L43
I'd expect the Rubygems to link to that version as well.

Reviewer 4: @p8 yeah. This is quite confusing. I wonder if there's some caching happening in that page?

Reviewer 6: @MatheusRich I think it needs a new release. The metadata was updated in a unreleased version.

Reviewer: Closing the loop here, this is now living under [rails/builder](https://github.com/rails/builder): https://github.com/rails/rails/commit/d462fb54b459e222efc2e2480397af44a9c55361

</div>
```

See [Builder documentation](https://github.com/jimweirich/builder) for more examples.

> ### Caching

Reviewer 6: The type of caching mentioned here (compiling to a method) is unrelated to the more commonly known Action View caching (storing a Template or a part of it in Redis for example). To avoid confusion I think we should rename this (and not mention Fragment Caching here).
```suggestion
### Template Compilation
```

By default, Rails will compile each template to a method to render it. In the development environment, when you alter a template, Rails will check the file's modification time and recompile it.

There is also Fragment Caching for when different parts of the page need to be cached and expired separately. Learn more about it in the [caching guide](caching_with_rails.html#fragment-caching).

Partials
--------

Partial templates - usually just called "partials" - are a way of breaking up the view templates into smaller reusable chunks. With partials, you can extract a piece of code from your main template to a separate smaller file, and render that file in the main template. You can also pass data to the partial files from the main template.

Let's see this in action with some examples:

### Rendering Partials

To render a partial as part of a view, you use the [`render`](https://api.rubyonrails.org/classes/ActionView/Helpers/RenderingHelper.html#method-i-render) method within the view:

```erb
<%= render "product" %>
```

> This will look for a file named `_product.html.erb` on the same folder to render within that view. Partial file names start with leading underscore character by convention. The file name distinguishes partials from regular views. However, no underscore is used when referring to partials for rendering within a view. This is true even when you reference a partial from another directory:

Reviewer 6: ```suggestion
This will look for a file named `_product.html.erb` in the same folder to render within that view. Partial file names start with leading underscore character by convention. The file name distinguishes partials from regular views. However, no underscore is used when referring to partials for rendering within a view. This is true even when you reference a partial from another directory:
```

<%= render "application/footer" %>
```

Here, the `_ad_banner.html.erb` and `_footer.html.erb` partials could contain content that is shared among many pages in your application. You don't need to see the details of these sections when you're focused on a Products' page.

The above example also uses the `_product.html.erb` partial. This partial contains details for rendering an individual product and is used to render each product in the collection `@products`.

> ### Passing Data to Partials With `locals` Option

Reviewer 6: ```suggestion
### Passing Data to Partials with `locals` Option
```

<% end %>
```

Another example from Active Storage's `_blob.html.erb`. This one sets the size based on whether `in_gallery` local variable is set when rendering the partial that contains this line:

```html+erb
<%= image_tag blob.representation(resize_to_limit: local_assigns[:in_gallery] ? [ 800, 600 ] : [ 1024, 768 ]) %>
```

### `render` without `partial` and `locals` Options

In the above examples, `render` takes 2 options: `partial` and `locals`. But if
these are the only options you need to use, you can skip the keys, `partial` and `locals` and specify the values only.

For example, instead of:

```erb
<%= render partial: "product", locals: { product: @product } %>
```

You can write:

```erb
<%= render "product", product: @product %>
```

You can also use this shorthand based on conventions:

```erb
<%= render @product %>
```

This will look for a partial named `_product.html.erb` in `app/views/products/`, as well as pass a local named `product` set to the value `@product`.

### The `as` and `object` Options

By default, objects passed to the template are in a local variable with the same name as the template. So, given:

```erb
<%= render @product %>
```

> within the `_product.html.erb` partial we'll get `@product` instance variable in the local variable `product`, as if we had written:

Reviewer 5: nit
It switches from "you" to "we" here. It'd be better to stay consistent throughout the document.

```erb
<%= render partial: "product", locals: { product: @product } %>
```

The `object` option can be used to specify a different name. This is useful when the template's object is elsewhere (e.g. in a different instance variable or in a local variable).

For example, instead of:

```erb
<%= render partial: "product", locals: { product: @item } %>
```

we can write:

```erb
<%= render partial: "product", object: @item %>
```

> This assigns a partial local variable named `product` to the instance variable `@item`. What if we wanted to change the local variable name from `product` by default to something else? We can use the `:as` option for that.

Reviewer 5: ```suggestion
This assigns the instance variable `@item` to a partial local variable named `product`. What if we wanted to change the local variable name from the default `product` to something else? We can use the `:as` option for that.
```

`locals: { my_product: @product, my_reviews: @reviews }`.

However, if a template refers to a variable that *isn't* passed into the view as
part of the `locals:` option, the template will raise an
`ActionView::Template::Error`:

```html+erb
<%# app/views/products/_product.html.erb %>

<%= tag.div id: dom_id(my_product) do %>
  <h1><%= my_product.name %></h1>

  <%# => raises ActionView::Template::Error for `product_reviews` %>
  <% product_reviews.each do |review| %>
    <%# ... %>
  <% end %>
<% end %>
```

### Using `local_assigns`

Each partial has a method called [local_assigns][] available. You can use this
method to access keys passed via the `locals:` option. If a partial was not
rendered with `:some_key` set, the value of `local_assigns[:some_key]` will be
`nil` within the partial.

For example, `product_reviews` is `nil` in the below example since only
`product` is set in `locals:`:

```html+erb
<%# app/views/products/show.html.erb %>

<%= render partial: "product", locals: { product: @product } %>

<%# app/views/products/_product.html.erb %>

<% local_assigns[:product]          # => "#<Product:0x0000000109ec5d10>" %>
<% local_assigns[:product_reviews]  # => nil %>
```

> One use case for `local_assigns` is optionally passing in a local variable and

Reviewer 4: Should we add a note that now this is also possibly using default values on strict locals?

Reviewer 4: That's enough. TY!

<%= render partial: @products, spacer_template: "product_ruler" %>
```

Rails will render the `_product_ruler.html.erb` partial (with no data passed to it) between each pair of `_product.html.erb` partials.

### Counter Variables

Rails also makes a counter variable available within a partial called by the collection. The variable is named after the title of the partial followed by `_counter`. For example, when rendering a collection `@products` the partial `_product.html.erb` can access the variable `product_counter`. The variable indexes the number of times the partial has been rendered within the enclosing view, starting with a value of `0` on the first render.

```erb
<%# index.html.erb %>
<%= render partial: "product", collection: @products %>
```

```erb
<%# _product.html.erb %>
<%= product_counter %> # 0 for the first product, 1 for the second product...
```

This also works when the local variable name is changed using the `as:` option. So if you did `as: :item`, the counter variable would be `item_counter`.

> Note: The following two sections, [Strick Locals](#strict-locals) and [Local Assigns with Pattern Matching](#local-assigns-with-pattern-matching) are more advance features of using partials, included here for completeness.

Reviewer 6: ```suggestion
Note: The following two sections, [Strict Locals](#strict-locals) and [Local Assigns with Pattern Matching](#local-assigns-with-pattern-matching) are more advanced features of using partials, included here for completeness.
```

product == record   # => true
```

You can also conditionally read a variable, then fall back to a default value when the key isn't part of the `locals:` options, using `fetch`:

```html+erb
<%# app/views/products/_product.html.erb %>

<% local_assigns.fetch(:related_products, []).each do |related_product| %>
  <%# ... %>
<% end %>
```

Combining Ruby 3.1's pattern matching assignment with calls to [Hash#with_defaults](https://api.rubyonrails.org/classes/Hash.html#method-i-with_defaults) enables compact partial-local default variable assignments:

```html+erb
<%# app/views/products/_product.html.erb %>

<% local_assigns.with_defaults(related_products: []) => { product:, related_products: } %>

<%= tag.div id: dom_id(product) do %>
  <h1><%= product.name %></h1>

  <% related_products.each do |related_product| %>
    <%# ... %>
  <% end %>
<% end %>
```

INFO: By default, partials will accept any `locals` as keyword arguments. To define what `locals` a partial accepts, use a `locals:` magic comment. To learn more, read about [Strict Locals](#strict-locals).

[local_assigns]: https://api.rubyonrails.org/classes/ActionView/Template.html#method-i-local_assigns

### Strict Locals

> Action View partials will accept any number of `locals` as keyword arguments. You can enforce how many and which `locals` a template accepts, set default value, and more with a `locals:` magic comment.

Reviewer 6: `locals` is plural here. Should `values` be plural as well?
```suggestion
Action View partials will accept any number of `locals` as keyword arguments. You can enforce how many and which `locals` a template accepts, set default values, and more with a `locals:` magic comment.
```

<%= message %>
```

Rendering the partial without a `:message` local variable uses the default value set in the `locals:` magic comment:

```ruby
render "messages/message"
# => "Hello, world!"
```

Rendering the partial with local variables not specified in the `local:` magic comment will also raise an exception:

```ruby
render "messages/message", unknown_local: "will raise"
# => ActionView::Template::Error: unknown local: :unknown_local for app/views/messages/_message.html.erb
```

You can allow optional local variable arguments with the double splat `**` operator:

```erb

<%# app/views/messages/_message.html.erb %>

<%# locals: (message: "Hello, world!", **attributes) -%>
<%= tag.p(message, **attributes) %>
```

> Or you can disable `locals` entirely by set the `locals:` to empty `()`:

Reviewer 6: ```suggestion
Or you can disable `locals` entirely by setting the `locals:` to empty `()`:
```

Reviewer 5: ```suggestion
Or you can disable `locals` entirely by setting `locals:` to `()`:
```

To find the layout for the current controller action, Rails first looks for a file in `app/views/layouts` with the same base name as the controller. For example, rendering actions from the PhotosController class will `use app/views/layouts/photos.html.erb`.

If such a controller-specific layout does not exist, Rails will use `app/views/layouts/application.html.erb`.

Here is an example of a simple layout in `application.html.erb` file:

```html+erb
<!DOCTYPE html>
<html>
<head>
  <title><%= "Your Rails App" %></title>
  <%= csrf_meta_tags %>
  <%= csp_meta_tag %>
  <%= stylesheet_link_tag "application", "data-turbo-track": "reload" %>
  <%= javascript_importmap_tags %>
</head>
<body>

<nav>
  <ul>
    <li><%= link_to "Home", root_path %></li>
    <li><%= link_to "Products", products_path %></li>
    <!-- Additional navigation links here -->
  </ul>
</nav>

<%= yield %>

<footer>
  <p>&copy; <%= Date.current.year %> Your Company</p>
</footer>
```

In the above example layout, view content will be rendered in place of `<%= yield %>`, and surrounded by the same `<head>`, `<nav>`, and `<footer>` content.

> Rails also provides more ways to assign specific layouts to individual controllers and actions. You can learn more about layouts in general in the [Layouts and Rendering in Rails](layouts_and_rendering.html) guide.

Reviewer 6: Tiny nitpick. `also` doesn't seem to add much here.
```suggestion
Rails provides more ways to assign specific layouts to individual controllers and actions. You can learn more about layouts in general in the [Layouts and Rendering in Rails](layouts_and_rendering.html) guide.
```

Supposing we use the same `_box` partial from above, this would produce the same output as the previous example.

### Collection with Partial Layouts

When rendering collections it is also possible to use the `:layout` option:

```erb
<%= render partial: "article", collection: @articles, layout: "special_layout" %>
```

> The layout will be rendered together with the partial for each item in the collection. The current object and object_counter variables, `article` and `article_count` in the above example, will be available in the layout as well, the same way they are within the partial.

Reviewer 6: ```suggestion
The layout will be rendered together with the partial for each item in the collection. The current object and object_counter variables, `article` and `article_counter` in the above example, will be available in the layout as well, the same way they are within the partial.
```

Localized Views
---------------

Action View has the ability to render different templates depending on the
current locale.

For example, suppose you have an `ArticlesController` with a show action. By
default, calling this action will render `app/views/articles/show.html.erb`. But

> if you set `I18n.locale = :de`, then `app/views/articles/show.de.html.erb` will
> be rendered instead. If the localized template isn't present, the undecorated
> version will be used. This means you're not required to provide localized views
> for all cases, but they will be preferred and used if available.

Reviewer: This made me pause for a moment, when we say "then .../show.de.html.erb will be rendered instead". We go on to explain what happens if the localized version is not present, but that first portion is the one that made me "wait".

Maybe we can word it to try to indicate it will "try" that localized first? Here's some food for thought suggestion, feel free to take it or drop it:

```suggestion
if you set `I18n.locale = :de`, then Action View will try to render the template
`app/views/articles/show.de.html.erb` first. If the localized template isn't
present, the undecorated version will be used. This means you're not required to
provide localized views for all cases, but they will be preferred and used if
available.
```
